demo-circuit/src/blaze_csw/constraints/data_structures.rs

```diff
@@ -0,0 +1,1838 @@
+    ) -> Result<ECPointSimulationGadget, SynthesisError> {
+        let one_minus_y_ed = te_pk_y_coordinate_g
+            .negate(cs.ns(|| "-y_ed"))?
+            .add_constant(cs.ns(|| "1 - y_ed"), &SimulatedFieldElement::one())?;
+
+        let one_plus_y_ed = te_pk_y_coordinate_g
+            .add_constant(cs.ns(|| "1 + y_ed"), &SimulatedFieldElement::one())?;
+
+        // Alloc x coordinate in SW representation
+        let sw_pk_x_coordinate_g =
+            NonNativeFieldGadget::<SimulatedFieldElement, FieldElement>::alloc(
+                cs.ns(|| "alloc x coord in SW"),
+                || {
+                    let a_over_three =
+                        <SimulatedCurveParameters as MontgomeryModelParameters>::COEFF_A
+                            * (SimulatedFieldElement::from(3u8)
+                                .inverse()
+                                .expect("Must be able to compute 3.inverse() in SimulatedField"));
+
+                    let b_inv = <SimulatedCurveParameters as MontgomeryModelParameters>::COEFF_B
+                        .inverse()
+                        .expect("B inverse must exist");
+
+                    let one_plus_te_y_coord = one_plus_y_ed.get_value().get()?;
+                    let one_minus_te_y_coord = one_minus_y_ed.get_value().get()?;
+
+                    let one_plus_te_y_coord_over_one_minus_te_y_coord = one_plus_te_y_coord
+                        * one_minus_te_y_coord.inverse().ok_or_else(|| {
+                            SynthesisError::Other(
+                                "Should be able to compute inverse of (1 - y_te)".to_string(),
+                            )
+                        })?;
+
+                    Ok((one_plus_te_y_coord_over_one_minus_te_y_coord + a_over_three) * b_inv)
+                },
+            )?;
+
+        // Check SW x coordinate
+        // x_sw * 3B(1 - y_ed) = 3(1 + y_ed) + A(1 - y_ed)
+        {
+            let three_times_one_plus_y_ed = one_plus_y_ed
+                .mul_by_constant(cs.ns(|| "3 * (1 + y_ed"), &SimulatedFieldElement::from(3u8))?;
+
+            let a_times_one_minus_y_ed = one_minus_y_ed.mul_by_constant(
+                cs.ns(|| "A * (1 - y_ed)"),
+                &<SimulatedCurveParameters as MontgomeryModelParameters>::COEFF_A,
+            )?;
+
+            let rhs = three_times_one_plus_y_ed.add(
+                cs.ns(|| "3(1 + y_ed) + A(1 - y_ed)"),
+                &a_times_one_minus_y_ed,
+            )?;
+
+            let three_b_times_one_minus_y_ed = one_minus_y_ed.mul_by_constant(
+                cs.ns(|| "3B * (1 - y_ed)"),
+                &(<SimulatedCurveParameters as MontgomeryModelParameters>::COEFF_B
+                    * SimulatedFieldElement::from(3u8)),
+            )?;
+
+            sw_pk_x_coordinate_g.mul_equals(
+                cs.ns(|| "check x_sw"),
+                &three_b_times_one_minus_y_ed,
+                &rhs,
+            )?;
+        }
```

Comment:
Although conversion does not take too many constraints in comparison to the scalar multiplication circuit, we can improve here using a single non-native product: Multiplying the transformation equality
```
	x_sw * (1 - y_ed) = 1/B (1 + y_ed) + A/3B (1 - y_ed)
```
by minus one yields
```
	x_sw * (y_ed - 1) = -1/B (y_ed + 1) + A/3B (y_ed - 1).
```
Replacing `y_ed + 1` by `y_ed - 1  + 2`, and gathering the `y_ed - 1` terms results in
the single product equality 
```
	(x_sw + 1/B - A/3B) * (y_ed - 1) = -2/B.
```

Reply:
Done